README.md
![workflow](https://github.com/baarsa/mss-auth/actions/workflows/ci.yml/badge.svg)

# Auth service for Opera Guide

Node.js service issuing JWT tokens with refresh token rotation.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.1] - 2021-11-22

### Fixed

-   Security issues

## [1.0.0] - 2021-11-09

### Added

-   Basic features

[Unreleased]: https://github.com/baarsa/mss-auth/compare/1.0.1...HEAD

[1.0.1]: https://github.com/baarsa/mss-auth/compare/1.0.0...1.0.1


app.test.js
const request = require('supertest');
const { createApp } = require('./app');
const { createMockRepo, createMockTokenGenerator } = require('./mocks');

let mockRepo;
let mockTokenGenerator;
let app;
// What is this test (unit, integration) and should we test changes/interaction with repo?

describe('app', () => {
    beforeEach(() => {
        mockRepo = createMockRepo();
        mockTokenGenerator = createMockTokenGenerator();
        app = createApp(mockRepo, mockTokenGenerator);
    });
    beforeAll(() => {
        expect.extend({
            toHaveCookie: (headers, cookieName) => {
                return {
                    message: () => `Expected ${headers} to contain cookie ${cookieName}`,
                    pass: headers['set-cookie'] !== undefined
                    && headers['set-cookie'].some(cookie => cookie.startsWith(cookieName)) };
            },
        });
    });
    describe('/signup', () => {
        it('should reject with 400 if request doesnt contain login', () => {
            return request(app)
                .post('/signup')
                .send({password: 'abcdef'})
                .set('Content-Type', 'application/json')
                .set('Accept', 'application/json')
                .expect(400);
        });
        // todo add more cases
        it('should return access token and set refreshToken cookie', () => {
            return request(app)
                .post('/signup')
                .send({login: 'bob_dylan', password: 'abcdef'})
                .expect(200)
                .then((response) => {
                    expect(response.body).toHaveProperty('token');
                    expect(response.headers).toHaveCookie('refreshToken');
                });
        });
    });
    describe('/login', () => {
        it('should reject with 401 for wrong credentials', () => {
            return request(app)
                .post('/login')
                .send({ login: 'Adrian', password: 'abcdefg'})
                .expect(401);
        });
        it('should return access token and set refreshToken cookie', () => {
            return request(app)
                .post('/login')
                .send({ login: 'Adrian', password: 'abcdef'})
                .expect(200)
                .then((response) => {
                    expect(response.body).toHaveProperty('token');
                    expect(response.headers).toHaveCookie('refreshToken');
                });
        });
    });
    describe('/logout', () => {
        it('should reject with 400 if request doesnt contain token', () => {
            return request(app)
                .post('/logout')
                .send({ })
                .expect(400);
        });
        it('should return 200 otherwise', () => {
            return request(app)
                .post('/logout')
                .set('Cookie', ['refreshToken=ok_rf'])
                .send()
                .expect(200);
        });
    });
    describe('/refresh', () => {
        it('should reject with 400 if request doesnt contain token', () => {
            return request(app)
                .post('/refresh')
                .send({ })
                .expect(400);
        });
        it('should reject with 401 if token family has been compromised', () => {
            return request(app)
                .post('/refresh')
                .set('Cookie', ['refreshToken=rf_of_compromised_family'])
                .send()
                .expect(401);
        });
        it('should reject with 401 if token has been used', () => {
            return request(app)
                .post('/refresh')
                .set('Cookie', ['refreshToken=used_rf'])
                .send()
                .expect(401);
        });
        it('should return new access and refresh tokens otherwise', () => {
            return request(app)
                .post('/refresh')
                .set('Cookie', ['refreshToken=ok_rf'])
                .send()
                .expect(200)
                .then((response) => {
                    expect(response.body).toHaveProperty('token');
                    expect(response.headers).toHaveCookie('refreshToken');
                });
        });
    });
    describe('/user', () => {
        it('should reject with 401 if no token', () => {
            return request(app)
                .get('/user')
                .expect(401);
        });
        it('should reject with 401 if token is invalid', () => {
            return request(app)
                .get('/user')
                .set({ Authorization: 'Bearer example-token-invalid' })
                .expect(401);
        });
        it('should return user name and role', () => {
            return request(app)
                .get('/user')
                .set({ Authorization: 'Bearer example-token-valid' })
                .expect(200)
                .then((response) => {
                    expect(response.body).toHaveProperty('name');
                    expect(response.body).toHaveProperty('role');
                });
        });
    });
});


mocks.js
const crs = require('crypto-random-string');

const createMockRepo = () => {
    const users = [
        { id: 1, login: 'Adrian', pass: '$2b$10$Vr4NNENrBAQCkUo4uyLMtOEOYhSr.kPhV9yE2ddUk2.20wKfgYf32', role: 1 },
        { id: 2, login: 'Benjamin', pass: '$2b$10$Vr4NNENrBAQCkUo4uyLMtOEOYhSr.kPhV9yE2ddUk2.20wKfgYf32', role: 1 }
    ];
    let currentUserId = 1;
    const refreshFamilies = [
        {
            id: 1,
            user: 1,
            has_been_compromised: false,
        },
        {
            id: 2,
            user: 2,
            has_been_compromised: true,
        }
    ];
    let currentFamilyId = 2;
    const refreshTokens = [
        {
            id: 1,
            value: 'ok_rf',
            family: 1,
            has_been_used: false,
        },
        {
            id: 2,
            value: 'rf_of_compromised_family',
            family: 2,
            has_been_used: false,
        },
        {
            id: 3,
            value: 'used_rf',
            family: 1,
            has_been_used: true,
        },
    ];
    let currentTokenId = 3;
    const mockRepo = {
        getUserById: (id) => users.find(user => user.id === id),
        getUserByName: (login) => users.find(user => user.login === login),
        getUserByNameAndPassword: (login, password) => users.find(user => user.login === login && user.password === password),
        addUser: (login, password) => {
            const newUser = {
                id: ++currentUserId,
                login,
                password, // hashed
                role: 2,
            };
            users.push(newUser);
            return newUser;
        },
        addRefreshFamily: (userId) => refreshFamilies.push({
            id: ++currentFamilyId,
            user: userId,
            has_been_compromised: false,
        }),
        addRefreshToken: (token, familyId) => refreshTokens.push({
            id: ++currentTokenId,
            value: token,
            family: familyId,
            has_been_used: false,
        }),
        getRefreshToken: (token) => refreshTokens.find(item => item.value === token),
        getRefreshFamily: (familyId) => refreshFamilies.find(item => item.id === familyId),
        markFamilyCompromised: (familyId) => refreshFamilies.find(item => item.id === familyId).compromised = true,
        markTokenUsed: (tokenId) => refreshTokens.find(item => item.id === tokenId).used = true,
        invalidateRefreshTokenFamiliesForUser: (userId) => refreshFamilies.find(item => item.user === userId).has_been_compromised = true,
    };
    return mockRepo;
};

const createMockTokenGenerator = () => {
    const generatedTokens = [];
    const mockTokenGenerator = {
        getTokens: (data) => {
            const newTokens = {
                token: crs({ length: 60 }),
                refreshToken: crs({ length: 60 }),
            };
            generatedTokens.push(newTokens);
            return newTokens;
        },
        verify: (token) => token === 'example-token-valid',
        decode: () => ({ name: 'Charlie', role: 'admin' })
    };
    return mockTokenGenerator;
};

module.exports = { createMockRepo, createMockTokenGenerator };


mysql-repo.js
// when and how connect/disconnect
const createRepo = (conn) => {
    const repo = {
        getUserById: async (id) => {
            const results = await conn.execute('SELECT u.id id, u.name name, u.pass pass, r.name role FROM user u JOIN role r ON r.id = u.role WHERE u.id=?', [id]);
            return results[0].length === 0 ? undefined : results[0][0];
        },
        getUserByName: async (login) => {
            const results = await conn.execute('SELECT u.id id, u.name name, u.pass pass, r.name role FROM user u JOIN role r ON r.id = u.role WHERE u.name=?', [login]);
            return results[0].length === 0 ? undefined : { id: results[0][0].id, name: results[0][0].name, pass: results[0][0].pass, role: results[0][0].role };
        },
        addUser: async (login, password) => {
            const newUser = [
                login,
                password, // hashed
                2,
            ];
            const results = await conn.execute('INSERT INTO user (name, pass, role) VALUES (?, ?, ?)', newUser);
            return (await repo.getUserById(results[0].insertId));
        },
        addRefreshFamily: async (userId) => {
            const newRefreshFamily = [
                userId,
                false,
            ];
            const results = await conn.execute('INSERT INTO refresh_family (user, has_been_compromised) VALUES (?, ?)', newRefreshFamily);
            return results[0].insertId;
        },
        addRefreshToken: async (token, familyId) => {
            const newRefreshToken = [
                token,
                familyId,
                false,
            ];
            return conn.execute('INSERT INTO refresh_token (value, family, has_been_used) VALUES (?, ?, ?)', newRefreshToken);
        },
        getRefreshToken: async (token) => {
            const results = await conn.execute('SELECT * FROM refresh_token WHERE value=?', [token]);
            return results[0].length === 0
                ? undefined
                : {
                    ...results[0][0],
                    has_been_used: results[0][0].has_been_used === 1,
                };
        },
        getRefreshFamily: async (familyId) => {
            const results = await conn.execute('SELECT * FROM refresh_family WHERE id=?', [familyId]);
            return results[0].length === 0
                ? undefined
                : {
                    ...results[0][0],
                    has_been_compromised: results[0][0].has_been_compromised === 1,
                };
        },
        markFamilyCompromised: async (familyId) => {
            return conn.execute('UPDATE refresh_family SET has_been_compromised=1 WHERE id=?', [familyId]);
        },
        markTokenUsed: async (tokenId) => {
            return conn.execute('UPDATE refresh_token SET has_been_used=1 WHERE id=?', [tokenId]);
        },
        invalidateRefreshTokenFamiliesForUser: async (id) => {
            return conn.execute('UPDATE refresh_family rf SET rf.has_been_compromised=1 WHERE rf.user=?', [id]);
        }
    };
    return repo;
};

module.exports = { createRepo };


app.js
const express = require('express');
const cookieParser = require('cookie-parser');
const bcrypt = require('bcrypt');

const ERROR_MESSAGE = 'Error connecting to the database';

const createApp = (repo, tokenManager) => {
    const app = express();
    app.use(cookieParser());
    app.use(express.json());
    app.post('/signup', async (req, res) => {
        // Check request parameters
        const body = req.body;
        if (body.login === undefined || body.password === undefined) {
            res.status(400);
            res.send();
            return;
        }
        // Else try to find user with this login from repo
        try {
            const user = await repo.getUserByName(body.login);
            // If he exists respond with 200 + message: 'login taken'
            if (user !== undefined) {
                res.status(200);
                res.send({ message: 'Login already taken' });
                return;
            }
            // Hash password
            const hash = bcrypt.hashSync(body.password, 10);
            // Add user to DB
            const newUser = await repo.addUser(body.login, hash);
            // Generate tokens
            const tokens = await tokenManager.getTokens({ name: newUser.name, role: newUser.role });
            // Add new refresh token and new family to DB
            const newRefreshFamilyId = await repo.addRefreshFamily(newUser.id);
            await repo.addRefreshToken(tokens.refreshToken, newRefreshFamilyId);
            // Send access token and set refresh token as a cookie
            res.cookie('refreshToken', tokens.refreshToken, { httpOnly: true, sameSite: 'strict' });
            res.send({
                token: tokens.token,
            });
        } catch (e) {
            res.status(500);
            res.send({ message: ERROR_MESSAGE });
        }
    });
    app.post('/login', async (req, res) => {
        // Check request parameters
        const { body } = req;
        if (body.login === undefined || body.password === undefined) {
            res.status(400);
            res.send();
            return;
        }
        // Get user record
        try {
            const user = await repo.getUserByName(body.login);
            if (user === undefined) {
                res.status(401);
                res.send();
                return;
            }
            // Verify password
            if (!bcrypt.compareSync(body.password, user.pass)) {
                res.status(401);
                res.send();
                return;
            }
            // Invalidate all previous refresh token families
            await repo.invalidateRefreshTokenFamiliesForUser(user.id);
            // Generate tokens
            const tokens = await tokenManager.getTokens({name: user.name, role: user.role});
            // Insert new token family,
            const newRefreshFamilyId = await repo.addRefreshFamily(user.id);
            // Insert new ref. token,
            await repo.addRefreshToken(tokens.refreshToken, newRefreshFamilyId);
            // Send access token and set refresh token as a cookie
            res.cookie('refreshToken', tokens.refreshToken, {httpOnly: true, sameSite: 'strict'});
            res.send({
                token: tokens.token,
            });
        } catch (e) {
            res.status(500);
            res.send({ message: ERROR_MESSAGE });
        }
    });
    app.post('/logout', async (req, res) => {
        // Request fields should include refreshToken
        const { cookies } = req;
        if (cookies.refreshToken === undefined) {
            res.status(400);
            res.send();
            return;
        }
        // Send 401 if token doesn't exist
        try {
            const token = await repo.getRefreshToken(cookies.refreshToken);
            if (token === undefined) {
                res.status(401);
                res.send();
                return;
            }
            // Send 401 if token family doesn't exist
            const tokenFamily = await repo.getRefreshFamily(token.family);
            if (tokenFamily === undefined) {
                res.status(401);
                res.send();
                return;
            }
            // Invalidate token family
            await repo.markFamilyCompromised(tokenFamily.id);
            // Drop refreshToken cookie and respond with success
            res.cookie('refreshToken', '');
            res.status(200);
            res.send();
        } catch (e) {
            res.status(500);
            res.send({ message: ERROR_MESSAGE });
        }
    })
    app.post('/refresh', async (req, res) => {
        // Check cookie
        const { cookies } = req;
        if (cookies.refreshToken === undefined) {
            res.status(400);
            res.send();
            return;
        }
        // Get refreshToken record
        try {
            const token = await repo.getRefreshToken(cookies.refreshToken);
            if (token === undefined) {
                res.status(401);
                res.send();
                return;
            }
            // Get token family
            const tokenFamily = await repo.getRefreshFamily(token.family);
            if (tokenFamily === undefined) {
                res.status(401);
                res.send();
                return;
            }
            // If family has been compromised respond with 401
            if (tokenFamily.has_been_compromised) {
                res.status(401);
                res.send();
                return;
            }
            // Else if token has been used
            // then mark family compromised and respond 401
            if (token.has_been_used) {
                await repo.markFamilyCompromised(tokenFamily.id);
                res.status(401);
                res.send();
                return;
            }
            // Else mark token used,
            await repo.markTokenUsed(token.id);
            const user = await repo.getUserById(tokenFamily.user);
            const tokens = await tokenManager.getTokens({name: user.name, role: user.role});
            // Generate new tokens
            // Insert refresh token with existing family
            await repo.addRefreshToken(tokens.refreshToken, tokenFamily.id);
            // Send access token and set refresh token as a cookie
            res.cookie('refreshToken', tokens.refreshToken, {httpOnly: true, sameSite: 'strict'});
            res.send({
                token: tokens.token,
            });
        } catch (e) {
            res.status(500);
            res.send({ message: ERROR_MESSAGE });
        }
    });
    app.get('/user', async (req, res) => {
        if (!req.headers.authorization || !req.headers.authorization.startsWith('Bearer ')) {
            res.status(401);
            res.send();
            return;
        }
        try {
            const token = req.headers.authorization.substr(7);
            if (!tokenManager.verify(token)) {
                res.status(401);
                res.send();
                return;
            }
            const data = tokenManager.decode(token);
            res.send(data);
        } catch (e) {
            res.status(500);
            res.send({ message: ERROR_MESSAGE });
        }
    });
    return app;
};

module.exports = { createApp };
